### ISSUE TITLE
Add purely negative images gets error when using object detection API to create tf records?

### ISSUE THREAD

**Issue by liangxiao05** [tensorflow/models#3578]
I met this problem when I tested my SSD model trained by object detection API.The model preforms well until it meets some images that contain the objects very simillar to the objective classfications.
I don't want to add another class to these wrong detected objects ,so I decide to include these images in my training datasets.
Refered to https://github.com/tensorflow/models/issues/2544，
''
You need to set min_negatives_per_image to a non zero number in the config for the model to sample boxes from purely negative images: 
''
After reviewed the code
https://github.com/tensorflow/models/blob/master/research/object_detection/core/losses.py#L433,
it tells me setting  'min_negatives_per_image ' to a positive number allows sampling negatives in an image without any positive anchors ,

However，if the image has no positive anchors , it contains no 'object' in its xml file ,then the script 
https://github.com/tensorflow/models/blob/master/research/object_detection/dataset_tools/create_pascal_tf_record.py#L107   will get this error when I creat the record :
for obj in data['object']:
   Key Error : 'object' 

I think many people are puzzled by this, please help us  to solve it .Hope for the replys ,thanks!

**Comment by twtygqyy** [comment]
Hi @liangxiao05 Could you please share your solution?

**Comment by marcoromelli** [comment]
@liangxiao05 can you share the solution you've found?

**Comment by liangxiao05** [comment]
@marcoromelli，sure!
First you can  set min_negatives_per_image to a non zero number in the config.
Then when you generate xmls , for purely negative images you can label their xmls only to contain the path,filename, width or height informations (for example ,if you generate the xml with labelImg,you don't have to draw a bbox on it)
then when creating tf records , take a 'if' judgement on it like this:
 -----
if data.has_key('object'):
   for obj in data['object']:
----
That's all ~~~~

**Comment by saiaman** [comment]
Do you need to change the other config variables in hard_example_miner ?

**Comment by leccyril** [comment]
I am sorry but i have same problem.. But i could not add negative picture because i could use algorithm for any picture or video,

I am trying to find only one object.. Then i test my detection on similar object, for example trained model with samsung gear s3 and detect in appel watch.. It says to me apple watch is a samsung watch... 

This is really annoying

There is no simple way to only detect object when is it found and not trying to detect object everywhere ? Even if not present at all ?

**Comment by jackweiwang** [comment]
Hi @liangxiao05  May I see your config file?

**Comment by AeroWRX** [comment]
> @marcoromelli，sure!
> First you can set min_negatives_per_image to a non zero number in the config.
> Then when you generate xmls , for purely negative images you can label their xmls only to contain the path,filename, width or height informations (for example ,if you generate the xml with labelImg,you don't have to draw a bbox on it)
> then when creating tf records , take a 'if' judgement on it like this:
> 
> ```
> if data.has_key('object'):
>     for obj in data['object']:
>     ... ...
> ```
> 
> That's all ~~~~

Can you explain which part of generate_tfrecord we have to modify with this?

**Comment by EricThomson** [comment]
> Hi everyone.
> I have the same Issue and I Notice that one file have no bounding box.
> i. e, looks like:
> 
> > 
> 
> ```
> <folder>Etiquetado rica11-part2</folder>
> <filename>18M10-14-15008-76530-008-1004873-1-1-4-16-2-1-1-21110.jpg</filename>
> <path>/home/usuario/Escritorio/PR/Etiquetado rica11-part2/18M10-14-15008-76530-008-1004873-1-1-4-16-2-1-1-21110.jpg</path>
> <source>
> 	<database>Unknown</database>
> </source>
> <size>
> 	<width>1812</width>
> 	<height>2416</height>
> 	<depth>3</depth>
> </size>
> <segmented>0</segmented>
> ```
> 
> The solution was to eliminate that file with its respective .jpg file and ejecute again the
> create_tf_record.py script.
> 
> I hope this will helpful

This excludes the negative example which is the opposite of what OP wants.

**Comment by EricThomson** [comment]
> > @marcoromelli，sure!
> > First you can set min_negatives_per_image to a non zero number in the config.
> > Then when you generate xmls , for purely negative images you can label their xmls only to contain the path,filename, width or height informations (for example ,if you generate the xml with labelImg,you don't have to draw a bbox on it)
> > then when creating tf records , take a 'if' judgement on it like this:
> > ```
> > if data.has_key('object'):
> >     for obj in data['object']:
> >     ... ...
> > ```
> > 
> > 
> > That's all ~~~~
> 
> Can you explain which part of generate_tfrecord we have to modify with this?

I made a github repo (tfrecord-view) with an example of how to create a tfrecord file, including hard negatives as above, for Python 3 (so no `has_key` method).  It also includes an example of how to consume a tfrecord file and display the images/bounding boxes with labels:
https://github.com/EricThomson/tfrecord-view

**Comment by AbdullahBahi** [comment]
> @marcoromelli，sure!
> First you can set min_negatives_per_image to a non zero number in the config.
> Then when you generate xmls , for purely negative images you can label their xmls only to contain the path,filename, width or height informations (for example ,if you generate the xml with labelImg,you don't have to draw a bbox on it)
> then when creating tf records , take a 'if' judgement on it like this:
> 
> ```
> if data.has_key('object'):
>     for obj in data['object']:
>     ... ...
> ```
> 
> That's all ~~~~

I think this works fine because it excludes all hard negatives (i.e. it doesn't create TFexamples for all XML files that doesn't contain any labeled objects which are the XMLs of the hard negatives)
so this actually terminates the error while creating the record files but it still doesn't solve the original issue (adding purely negative images to training records)

**Comment by AbdullahBahi** [comment]
> > > @marcoromelli，sure!
> > > First you can set min_negatives_per_image to a non zero number in the config.
> > > Then when you generate xmls , for purely negative images you can label their xmls only to contain the path,filename, width or height informations (for example ,if you generate the xml with labelImg,you don't have to draw a bbox on it)
> > > then when creating tf records , take a 'if' judgement on it like this:
> > > ```
> > > if data.has_key('object'):
> > >     for obj in data['object']:
> > >     ... ...
> > > ```
> > > 
> > > 
> > > That's all ~~~~
> > 
> > 
> > Can you explain which part of generate_tfrecord we have to modify with this?
> 
> I made a github repo (tfrecord-view) with an example of how to create a tfrecord file, including hard negatives as above, for Python 3 (so no `has_key` method). It also includes an example of how to consume a tfrecord file and display the images/bounding boxes with labels:
> https://github.com/EricThomson/tfrecord-view

can you please tell me how you created TFExample for negative samples -by negative here i mean an XML file with image info only(width,height and name) with no labeled objects

**Comment by AbdullahBahi** [comment]
@EricThomson can you explain this part for me please? 
`else:
        if verbose: print("No annotations for this one")
        obj_features = {
                'image/height': int64_feature(height),
                'image/width': int64_feature(width),
                'image/filename': bytes_feature(data['filename'].encode('utf8')),
                'image/encoded': bytes_feature(encoded_image),
                'image/annotated': int64_feature(0)
        }
    
    tf_features = tf.train.Features(feature = obj_features)
    tf_example = tf.train.Example(features = tf_features)`

**Comment by cortical-iv** [comment]
> 
> 
> > > > @marcoromelli，sure!
> > > > First you can set min_negatives_per_image to a non zero number in the config.
> > > > Then when you generate xmls , for purely negative images you can label their xmls only to contain the path,filename, width or height informations (for example ,if you generate the xml with labelImg,you don't have to draw a bbox on it)
> > > > then when creating tf records , take a 'if' judgement on it like this:
> > > > ```
> > > > if data.has_key('object'):
> > > >     for obj in data['object']:
> > > >     ... ...
> > > > ```
> > > > 
> > > > 
> > > > That's all ~~~~
> > > 
> > > 
> > > Can you explain which part of generate_tfrecord we have to modify with this?
> > 
> > 
> > I made a github repo (tfrecord-view) with an example of how to create a tfrecord file, including hard negatives as above, for Python 3 (so no `has_key` method). It also includes an example of how to consume a tfrecord file and display the images/bounding boxes with labels:
> > https://github.com/EricThomson/tfrecord-view
> 
> can you please tell me how you created TFExample for negative samples -by negative here i mean an XML file with image info only(width,height and name) with no labeled objects

I made them using labelimg: just go to the image and click 'verify image' but don't put any bounding boxes in that image, and save the xml file.

**Comment by EricThomson** [comment]
> 
> 
> > > > @marcoromelli，sure!
> > > > First you can set min_negatives_per_image to a non zero number in the config.
> > > > Then when you generate xmls , for purely negative images you can label their xmls only to contain the path,filename, width or height informations (for example ,if you generate the xml with labelImg,you don't have to draw a bbox on it)
> > > > then when creating tf records , take a 'if' judgement on it like this:
> > > > ```
> > > > if data.has_key('object'):
> > > >     for obj in data['object']:
> > > >     ... ...
> > > > ```
> > > > 
> > > > 
> > > > That's all ~~~~
> > > 
> > > 
> > > Can you explain which part of generate_tfrecord we have to modify with this?
> > 
> > 
> > I made a github repo (tfrecord-view) with an example of how to create a tfrecord file, including hard negatives as above, for Python 3 (so no `has_key` method). It also includes an example of how to consume a tfrecord file and display the images/bounding boxes with labels:
> > https://github.com/EricThomson/tfrecord-view
> 
> can you please tell me how you created TFExample for negative samples -by negative here i mean an XML file with image info only(width,height and name) with no labeled objects

I used imgaug: go to the image and click 'verify image' and save it (with no bounding box). It saves the xml file with no bounding box information but the rest of the metadata is there.

**Comment by EricThomson** [comment]
> 
> 
> @EricThomson can you explain this part for me please?
> 
> `if verbose: print("No annotations for this one") obj_features = { 'image/height': int64_feature(height), 'image/width': int64_feature(width), 'image/filename': bytes_feature(data['filename'].encode('utf8')), 'image/encoded': bytes_feature(encoded_image), 'image/annotated': int64_feature(0) } tf_features = tf.train.Features(feature = obj_features) tf_example = tf.train.Example(features = tf_features)`

That is formatted such as to make it unreadable to me. I would recommend going to the repo and opening an issue, this is probably not the place to discuss my repo. :)

**Comment by sampathkumaran90** [comment]
Hi, I am facing similar issue. When i set min_negatives_per_image to 0 and not set any negative images, i get prediction correct for positive images but getting a lot of false positives. When i add a bunch of negative images say 50% positive and 50% negative and set min_negatives_per_image to 10, the results are not good. I am getting more false negatives as well as false positives. Anyone could help me on this issue... I am using ssd mobilenet v2 or finetuning
